### PR TITLE
[Deepseek V4] Dsv4 sparse indexer loss

### PR DESCRIPTION
**Pull request by Matrix-Z97**
## Summary

Add DeepSeek V4 sparse indexer auxiliary loss for CSA layers.

This change wires the community `AuxLoss` framework into DeepSeek V4 compressed sparse attention so the sparse indexer can be trained from the main attention distribution. The loss is implemented as `SparseIndexerLoss` under `deepseek_v4/compressor.py` and is configured only for `compress_ratio == 4` CSA layers.

## Changes

- Add `SparseIndexerLoss(AuxLoss)` for indexer distillation.
- Update `Indexer.select(...)` to return both:
  - `topk_indices`
  - differentiable `topk_scores`
- Inject indexer loss through FlexAttention `out_transform`, using attention LSE as teacher normalization.
- Keep teacher tensors detached while preserving gradient through indexer scores.
- Register `register_aux_loss_zero_hook` in the DeepSeek V4 optimizer post-build hook.
- Add unit coverage for loss value/gradient/metric accumulation, DSV4 config wiring, and `Indexer.select(...)`.

## Validation

- `python -m py_compile torchtitan/models/deepseek_v4/compressor.py torchtitan/models/deepseek_v4/attention.py tests/unit_tests/cpu/test_deepseek_v4_indexer_loss.py`
- 4-GPU smoke test with:
  - `tp=2`
  - `ep=2`
  - DeepSeek V4 `spmd_types` path
  - training completed successfully
- Probe confirmed:
  - `Indexer.select(...)` returned `topk_scores` with `requires_grad=True`
  - `SparseIndexerLoss.forward(...)` received differentiable scores
  - aux loss accumulation was non-zero

## Notes

Local pytest collection is blocked by the local environment missing `torch_remat`; this is an environment dependency issue and matches the existing local setup behavior.

### CODE REVIEW

**Comment by tianyu-l** [comment]
@claude review this PR

Also, would it work well with optimizations in https://github.com/pytorch/torchtitan/pull/4452?

**Comment by drisspg** [comment]
I am about to put up a pr to update our impls to attn-gym, can we wait till after this change to land this?
